### PR TITLE
Suppress output of CLI spinner if not connected to terminal

### LINKED ISSUE #1066: Suppress sending ansi code if the output is not attached to TTY
**Bug description**

When you use tanzu cli in combination with `watch` it spits out a bunch of ansi control characters that aren't supported in that context.

To reproduce do something like this in a terminal:

```
watch tanzu package installed list -A
```

You'll get output that looks like this:

```
Every 2.0s: tanzu package installed list -A                                                                                           dellicioso: Thu Nov  4 13:32:11 2021

^[K | Retrieving installed packages... ^[K / Retrieving installed packages... ^[K - Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrievi
ng installed packages... ^[K / Retrieving installed packages... ^[K - Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed p
ackages... ^[K / Retrieving installed packages... ^[K - Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed packages... ^[K
 / Retrieving installed packages... ^[K - Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed packages... ^[K / Retrieving
installed packages... ^[K - Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed packages... ^[K / Retrieving installed pack
ages... ^[K - Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed packages... ^[K / Retrieving installed packages... ^[K -
Retrieving installed packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed packages... ^[K / Retrieving installed packages... ^[K - Retrieving ins
talled packages... ^[K \ Retrieving installed packages... ^[K | Retrieving installed packages... ^[K / Retrieving installed packages... ^[K - Retrieving installed package
s... ^[K \ Retrieving installed packages... ^[K
  ^[1mNAME                              ^[0m  ^[1mPACKAGE-NAME                                       ^[0m  ^[1mPACKAGE-VERSION^[0m  ^[1mSTATUS             ^[0m  ^[1mNAMES
PACE  ^[0m
  appliveview                         appliveview.tanzu.vmware.com                         0.3.0-build6     Reconcile succeeded  tap-install
  buildservice                        buildservice.tanzu.vmware.com                        1.3.1            Reconciling          tap-install
  cartographer                        cartographer.tanzu.vmware.com                        0.0.7            Reconcile succeeded  tap-install
  cnrs                                cnrs.tanzu.vmware.com                                1.0.3            Reconcile succeeded  tap-install
  conventions-controller              controller.conventions.apps.tanzu.vmware.com         0.4.2            Reconcile succeeded  tap-install
  developer-conventions               developer-conventions.tanzu.vmware.com               0.3.0-build.1    Reconcile succeeded  tap-install
  image-policy-webhook                image-policy-webhook.signing.run.tanzu.vmware.com    1.0.0-beta.1     Reconcile succeeded  tap-install
  ootb-supply-chain-basic             ootb-supply-chain-basic.tanzu.vmware.com             0.3.0-build.4    Reconcile succeeded  tap-install
  ootb-supply-chain-testing           ootb-supply-chain-testing.tanzu.vmware.com           0.3.0-build.4    Reconcile succeeded  tap-install
  ootb-supply-chain-testing-scanning  ootb-supply-chain-testing-scanning.tanzu.vmware.com  0.3.0-build.4    Reconcile succeeded  tap-install
  ootb-templates                      ootb-templates.tanzu.vmware.com                      0.3.0-build.4    Reconcile succeeded  tap-install
  service-bindings                    service-bindings.labs.vmware.com                     0.5.0            Reconcile succeeded  tap-install
  services-toolkit                    services-toolkit.tanzu.vmware.com                    0.4.0-rc.2       Reconcile succeeded  tap-install
  source-controller                   controller.source.apps.tanzu.vmware.com              0.1.2            Reconcile succeeded  tap-install
  spring-boot-conventions             spring-boot-conventions.tanzu.vmware.com             0.1.2            Reconcile succeeded  tap-install
  tap                                 tap.tanzu.vmware.com                                 0.3.0-build.6    Reconciling          tap-install
  tap-gui                             tap-gui.tanzu.vmware.com                             0.3.0-rc.4       Reconcile succeeded  tap-install
```

Notice the cruft at the beginning. 

This is a minor nuisance, but it would be good to fix it. (I.e. detect when the output isn't attached directly to a terminal and refrain in that case from sending ansi codes to do 'fancy' things that only a terminal understands).

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [ *] CLI
- [ ] Docs
- [ ] IAM
- [* ] Installation
- [ *] Plugin
- [ ] Security
- [ ] Test and Release
- [ *] User Experience

**Expected behavior**

Suppress sending ansi code if the output is not attached to TTY.

**Version** (include the SHA if the version is not obvious)

```
$ tanzu version
version: v0.10.0
buildDate: 2021-11-03
sha: fd96bebe
```

```
$ tanzu plugin list
  NAME                LATEST VERSION  DESCRIPTION                                                        REPOSITORY  VERSION  STATUS     
  apps                                Applications on Kubernetes                                                     v0.2.0   installed  
  cluster             v0.10.0         Kubernetes cluster operations                                      core        v0.10.0  installed  
  kubernetes-release  v0.10.0         Kubernetes release operations                                      core        v0.10.0  installed  
  login               v0.10.0         Login to the platform                                              core        v0.10.0  installed  
  management-cluster  v0.10.0         Kubernetes management cluster operations                           core        v0.10.0  installed  
  package             v0.10.0         Tanzu package management                                           core        v0.10.0  installed  
  pinniped-auth       v0.10.0         Pinniped authentication operations (usually not directly invoked)  core        v0.10.0  installed  
  secret              v0.10.0         Tanzu secret management                                            core        v0.10.0  installed  

```

**Environment where the bug was observed (cloud, OS, etc)**

Ubuntu 18.04, bash terminal

### PR DESCRIPTION
**Pull request by blc1996**
Signed-off-by: Lucheng Bao <luchengb@vmware.com>

### What this PR does / why we need it

CLI spinner uses ANSI control code to do things like erase the current line. This control code will only be recognized by terminal.

Disable the spinner if the data is piped.

### Which issue(s) this PR fixes

Fixes #1066 #1016 

### Describe testing done for PR
test pipelines
Manual testing on local machine

### Release note

```release-note
None
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by ggpaue** [comment]
@blc1996 change looks good but I was not able to reproduce this issue, are you able to reproduce?

**Comment by blc1996** [comment]
> @blc1996 change looks good but I was not able to reproduce this issue, are you able to reproduce?

Triggered package plugin test to verify https://github.com/vmware-tanzu/tanzu-framework/runs/4158455918?check_suite_focus=true

**Comment by maralavi** [comment]
This change solves the issue for Repository list/get and Secret operations and Package Available list/get and Installed list/get; but doesn't solve the issue for Package install/delete/update and Repository add/delete/update.

**Comment by blc1996** [comment]
> This change solves the issue for Repository list/get and Secret operations and Package Available list/get and Installed list/get; but doesn't solve the issue for Package install/delete/update and Repository add/delete/update.

Pushed new changes to take care of the rest of the commands. Let's wait for the pipeline results.

**Comment by maralavi** [comment]
Thanks for the changes Lucheng, they look good. Yes, lets wait for the PR pipeline

**Review by maralavi — APPROVED** [review]

**Comment by maralavi** [comment]
Ah you are getting the same error that Daniel was mentioning the other day. The following private repository gets added and reconciled without the secret being there. This instability can be coming from the private repository bundle that it can be added sometimes without the existence of the secret. When I tried to add this repository locally last week, it never reconciled even after the secret was added (was getting reconciled locally a month ago). One option to unblock the PR might be to comment out this private repository part of the test for later.

```
tanzu package repository update carvel-test --url projects-stg.registry.vmware.com/tkgprivate/example-pkg-repo@sha256:a80e9b512b9eff76ab638cce50a3c4541a12673d9b698103314f32c93f1deb61 --namespace test-ns --create-namespace --create --wait=true --poll-interval 20s --poll-timeout 20s --kubeconfig /home/tanzu/.kube-tkg/config 
```

**Comment by blc1996** [comment]
@ggpaue @vijaykatam can I get a review on this?

File: cmd/cli/plugin/package/display_utils.go [hunk 1]
```diff
@@ -14,15 +15,15 @@ import (
...
+func DisplayProgress(initialMsg string, pp *tkgpackagedatamodel.PackageProgress) error { //nolint:gocyclo
 	var (
 		currMsg string
 		s       *spinner.Spinner
 		err     error
 	)
 
 	newSpinner := func() (*spinner.Spinner, error) {
-		s = spinner.New(spinner.CharSets[9], 100*time.Millisecond)
+		s = spinner.New(spinner.CharSets[9], 100*time.Millisecond, spinner.WithWriter(os.Stderr))
```
Comment (vijaykatam) [inline cmd/cli/plugin/package/display_utils.go:26]: Why are we replacing [default colorized stdout](https://github.com/fatih/color/blob/master/color.go#L27) with os.Stderr?

File: cmd/cli/plugin/package/display_utils.go [hunk 2]
```diff
@@ -44,10 +47,17 @@ func DisplayProgress(initialMsg string, pp *tkgpackagedatamodel.PackageProgress)
 	}
 
 	s.Suffix = fmt.Sprintf(" %s", initialMsg)
-	s.Start()
 
+	fi, _ := os.Stdout.Stat()
+	// Start the spinner only if connected to terminal
```
Comment (vijaykatam) [inline cmd/cli/plugin/package/display_utils.go:52]: The comment describes what we should do but I think could we look at using a dependency like https://github.com/mattn/go-isatty or similar which is more comprehensive.

File: cmd/cli/plugin/package/display_utils.go (hunk 1 above)
Comment (blc1996) [reply cmd/cli/plugin/package/display_utils.go:26]: Tried to follow the docs of the spinner component. https://github.com/briandowns/spinner#unix-pipe-and-redirect

File: cmd/cli/plugin/package/display_utils.go (hunk 2 above)
Comment (blc1996) [reply cmd/cli/plugin/package/display_utils.go:52]: Updated.

File: cmd/cli/plugin/package/display_utils.go (hunk 1 above)
Comment (vijaykatam) [reply cmd/cli/plugin/package/display_utils.go:27]: If I am reading the doc correctly this is recommended for chain, pipe or redirecting the output. Could we verify if this is still needed with tty check?

**Review by vijaykatam — APPROVED** [review]
LGTM, thanks for testing alternate library.

**Comment by vijaykatam** [comment]
Lint is failing

**Pull request merged**